File: src/podcast/services/research.py
import logging
import time
import uuid

from podcast.database import get_session
from podcast.models import Episode
from podcast.services.llm_providers import complete, get_research_model

logger = logging.getLogger(__name__)

RESEARCH_SYSTEM_PROMPT_TEMPLATE = """You are a podcast research assistant. Given a topic, produce comprehensive \
research notes that will be used to write a podcast episode transcript.

Your research should include:
- Key facts and background information
- Interesting angles and perspectives
- Recent developments and current state
- Common misconceptions or surprising findings
- Potential discussion points and debate areas
- Relevant examples, case studies, or anecdotes

Be thorough but organized. Use clear headings and bullet points. \
The research should provide enough material for a {duration_description} conversational podcast episode."""

# Research config scaled by target episode length
RESEARCH_LENGTH_CONFIG = {
    15: {"duration_description": "10-15 minute", "max_tokens": 4096},
    30: {"duration_description": "25-30 minute", "max_tokens": 8192},
    60: {"duration_description": "55-60 minute", "max_tokens": 12000},
    120: {"duration_description": "2-hour", "max_tokens": 16000},
}


async def run_research(episode_id: uuid.UUID) -> dict:
    """Research a topic using the configured LLM provider. Returns metrics dict."""
    # Read episode data
    async with get_session() as db:
        episode = await db.get(Episode, episode_id)
        if not episode:
            raise ValueError(f"Episode {episode_id} not found")
        topic = episode.topic
        model_key = episode.research_model
        target_length = episode.target_length_minutes

    model_info = get_research_model(model_key)

    config = RESEARCH_LENGTH_CONFIG.get(target_length, RESEARCH_LENGTH_CONFIG[30])
    system_prompt = RESEARCH_SYSTEM_PROMPT_TEMPLATE.format(
        duration_description=config["duration_description"],
    )

    logger.info(
        "Researching topic for episode %s via %s (%s): %s",
        episode_id,
        model_info.display_name,
        model_info.model_id,
        topic[:100],
    )

    t0 = time.monotonic()
    response = await complete(
        model_info,
        system=system_prompt,
        user_message=f"Research the following topic thoroughly:\n\n{topic}",
        max_tokens=config["max_tokens"],
        use_web_search=True,
    )
    duration = time.monotonic() - t0

    if not response.text:
        raise RuntimeError("No research content generated")

    # Save results
    async with get_session() as db:
        episode = await db.get(Episode, episode_id)
        episode.research_notes = response.text

    metrics = {
        "model": response.model,
        "provider": model_info.provider,
        "input_tokens": response.input_tokens,
        "output_tokens": response.output_tokens,
        "duration_seconds": round(duration, 2),
        "output_chars": len(response.text),
    }
    logger.info(
        "Research complete for episode %s (%d chars, %d in/%d out tokens, %.1fs)",
        episode_id,
        len(response.text),
        metrics["input_tokens"],
        metrics["output_tokens"],
        duration,
    )
    return metrics

File: src/services/research.ts
import { eq } from "drizzle-orm";
import { db } from "../database.js";
import { episodes } from "../schema.js";
import { complete, getResearchModel } from "./llm-providers.js";
import { createLogger } from "../log-handler.js";

const log = createLogger("research");

const RESEARCH_SYSTEM_PROMPT_TEMPLATE = `You are a podcast research assistant. Given a topic, produce comprehensive \
research notes that will be used to write a podcast episode transcript.

Your research should include:
- Key facts and background information
- Interesting angles and perspectives
- Recent developments and current state
- Common misconceptions or surprising findings
- Potential discussion points and debate areas
- Relevant examples, case studies, or anecdotes

Be thorough but organized. Use clear headings and bullet points. \
The research should provide enough material for a {duration_description} conversational podcast episode.`;

const RESEARCH_LENGTH_CONFIG: Record<
  number,
  { durationDescription: string; maxTokens: number }
> = {
  15: { durationDescription: "10-15 minute", maxTokens: 4096 },
  30: { durationDescription: "25-30 minute", maxTokens: 8192 },
  60: { durationDescription: "55-60 minute", maxTokens: 12000 },
  120: { durationDescription: "2-hour", maxTokens: 16000 },
};

export async function runResearch(
  episodeId: string
): Promise<Record<string, unknown>> {
  // Read episode data
  const [episode] = await db
    .select()
    .from(episodes)
    .where(eq(episodes.id, episodeId))
    .limit(1);
  if (!episode) throw new Error(`Episode ${episodeId} not found`);

  const modelInfo = getResearchModel(episode.researchModel);
  const config = RESEARCH_LENGTH_CONFIG[episode.targetLengthMinutes] ??
    RESEARCH_LENGTH_CONFIG[30];

  const systemPrompt = RESEARCH_SYSTEM_PROMPT_TEMPLATE.replace(
    "{duration_description}",
    config.durationDescription
  );

  log.info(
    "Researching topic for episode %s via %s (%s): %s",
    episodeId,
    modelInfo.displayName,
    modelInfo.modelId,
    episode.topic.slice(0, 100)
  );

  const MAX_EMPTY_RETRIES = 2;
  const t0 = performance.now();
  let response = null;

  for (let attempt = 0; attempt <= MAX_EMPTY_RETRIES; attempt++) {
    response = await complete(
      modelInfo,
      systemPrompt,
      `Research the following topic thoroughly:\n\n${episode.topic}`,
      config.maxTokens,
      1.0,
      true
    );
    if (response.text) break;
    if (attempt < MAX_EMPTY_RETRIES) {
      const wait = 3 * 2 ** attempt;
      log.warn(
        "Research for episode %s returned empty text (attempt %d/%d, model=%s), retrying in %ds...",
        episodeId,
        attempt + 1,
        MAX_EMPTY_RETRIES + 1,
        modelInfo.modelId,
        wait
      );
      await new Promise((r) => setTimeout(r, wait * 1000));
    }
  }

  const duration = (performance.now() - t0) / 1000;

  if (!response?.text) {
    throw new Error(
      `No research content generated after ${MAX_EMPTY_RETRIES + 1} attempts ` +
        `(model=${modelInfo.modelId}). The API returned 200 OK but the response contained no text output.`
    );
  }

  // Save results
  await db
    .update(episodes)
    .set({ researchNotes: response.text })
    .where(eq(episodes.id, episodeId));

  const metrics = {
    model: response.model,
    provider: modelInfo.provider,
    input_tokens: response.inputTokens,
    output_tokens: response.outputTokens,
    duration_seconds: Math.round(duration * 100) / 100,
    output_chars: response.text.length,
  };

  log.info(
    "Research complete for episode %s (%d chars, %d in/%d out tokens, %.1fs)",
    episodeId,
    response.text.length,
    metrics.input_tokens,
    metrics.output_tokens,
    duration
  );

  return metrics;
}
